wgsl/index.bs

```diff
@@ -6981,13 +6979,296 @@ See [[#statements]] and [[#function-calls]].
+
+        To express that something must always be uniform (e.g. the control flow at the call site of a derivative), we simply add an edge from MustBeUniform to the corresponding node.
+        One way to understand this, is that MustBeUniform corresponds to the proposition True, so that MustBeUniform -> X is the same as saying that X is true.
+
+        Reciprocally, to express that we cannot ensure the uniformity of something (e.g. a variable which holds the thread id), we simply add an edge from the corresponding node to CannotBeUniform.
+        One way to understand this, is that CannotBeUniform corresponds to the proposition False, so that X -> CannotBeUniform is the same as saying that X is false.
+
+        A consequence of this interpretation is that every node reachable from MustBeUniform corresponds to something which must be uniform for the program to be valid, and every node from which cannotBeUniform is reachable corresponds to something whose uniformity we cannot guarantee. It follows that we have a uniformity violation (and thus reject the program) if there is any path from MustBeUniform to CannotBeUniform.
+</div>
+
+In more detail, here is what we compute for each function:
+
+* A tag about the control-flow in which the function can be called: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter}
+* A tag for each parameter of the function: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter, MustBeUniformForReturnValue, NoRestriction}
+* A tag for the function as a whole: one of {ControlFlowCannotBeUniformAfter, ReturnValueCannotBeUniform, NoRestriction}
+
+Here is the algorithm for computing this data for a given function:
+
+* Create nodes called "MustBeUniform", "CannotBeUniform", "CF_start", "CF_return", and if the function is non-void a node called "Value_return"
+* Create one node for each parameter of the function which we'll call "arg_i"
+* Walk over the syntax of the function, adding nodes and edges to the graph following the rules of the next section.
+* Look at which nodes are reachable from "MustBeUniform"
+    * If this set includes the node "CannotBeUniform", then reject the program.
+    * If this set includes "CF_start", then the control-flow in which the function can be called is AlwaysMustBeUniform.
+    * Otherwise it is MustBeUniformForControlFlowAfter
+    * For each "arg_i" in this set, the corresponding parameter is AlwaysMustBeUniform
+    * Remove from the graph all nodes that have been visited
+* Look at which nodes are reachable from "CF_return"
+    * If this set includes "CannotBeUniform", then the function is ControlFlowCannotBeUniformAfter
+    * For each "arg_i" in this set, the corresponding parameter is MustBeUniformForControlFlowAfter
+    * Remove from the graph all nodes that have been visited
+* If "Value_return" exists, look at which nodes are reachable from it
```

Comment:
The priority between Value_return and CF_return seems backward.

As proposed:
* If arg_i is reachable from CF_return, then arg_i is MustBeUniformForControlFlowAfter
    *  then remove arg_i
* If arg_i survives the previous step, **and** arg_i is reachable from Value_return, then arg_i is set to MustBeUniformForReturnValue
* But MustBeUniformForReturnValue is a strictly stronger requirement than MustBeUniformForControlFlowAfter.  That’s because an expression E being uniform requires E to be evaluated in uniform control flow **and** all invocations compute the same value.

Reply:
Let's consider the following example:
```
fn foo(a: bool, b: i32) -> i32 {
  if (a) {
    discard;
  }
  return b;
}
```
I believe that the rules I wrote would result in a path from `CF_return` to `arg_a`, and paths from `Value_return` to `arg_a` and `arg_b` (from running them on paper, I may have done a mistake). So `a` would be `MustBeUniformForControlFlowAfter`, and b would be `MustBeUniformForReturnValue`.

I believe this is correct, thanks to the following sentence in the rule for function calls:
> Add an edge from Result to CF_after

a call `foo(x, y)` would give the following graph: `CF_after -> {x, CF_before}, Result -> {y, CF_after}`.
Which seems fine to me: there is a path from Result to both `x` and `y` (so both are required to be uniform for it to be), and there is only a path from `CF_after` to `x` (so y can be non-uniform without affecting the uniformity of the control-flow).

Reply:
Got it.
I had not read as far as the dependency graph construction rules for function calls, and now see that things are wired up correctly there. 

thanks for the explanation.